server/commands/connected_objects.go

```diff
@@ -0,0 +1,351 @@
+	if req.Limit > 0 {
+		foundCount = new(uint32)
+	}
+
+	var foundObjects sync.Map
+	return c.streamedConnectedObjects(ctx, req, resultChan, &foundObjects, foundCount)
+}
+
+type reverseExpandRequest struct {
+	storeID          string
+	ingress          *graph.RelationshipIngress
+	sourceObjectRef  *openfgapb.RelationReference
+	targetUserRef    *openfgapb.ObjectRelation
+	contextualTuples []*openfgapb.TupleKey
+	limit            uint32
+}
+
+func (c *ConnectedObjectsCommand) reverseExpandTupleToUserset(
+	ctx context.Context,
+	req *reverseExpandRequest,
+	resultChan chan<- string,
+	foundObjectsMap *sync.Map,
+	foundCount *uint32,
+) error {
+
+	store := req.storeID
+
+	ingress := req.ingress.Ingress
+
+	sourceObjectType := req.sourceObjectRef.GetType()
+	sourceObjectRel := req.sourceObjectRef.GetRelation()
+
+	tuplesetRelation := req.ingress.TuplesetRelation.GetRelation()
+
+	var tuples []*openfgapb.Tuple
+	for _, t := range req.contextualTuples {
+
+		object := t.GetObject()
+		objectType, _ := tuple.SplitObject(object)
+		if objectType != ingress.GetType() {
+			continue
+		}
+
+		if t.GetRelation() != tuplesetRelation {
+			continue
+		}
+
+		targetUserStr := req.targetUserRef.GetObject()
+
+		userObj, _ := tuple.SplitObjectRelation(t.GetUser())
+		if userObj == targetUserStr || userObj == "*" {
+			tuples = append(tuples, &openfgapb.Tuple{Key: t})
+		}
+	}
+	iter1 := storage.NewStaticTupleIterator(tuples)
+
+	iter2, err := c.Datastore.ReadStartingWithUser(ctx, store, storage.ReadStartingWithUserFilter{
+		ObjectType: req.ingress.Ingress.GetType(),
+		Relation:   tuplesetRelation,
+		UserFilter: []*openfgapb.ObjectRelation{
+			{Object: req.targetUserRef.Object},
+			{Object: "*"},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	iter := storage.NewCombinedIterator(iter1, iter2)
+
+	g := errgroup.Group{}
+	g.SetLimit(100) // set some concurrency limit
+
+	for {
+		t, err := iter.Next()
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
+
+			return err
+		}
+
+		tk := t.GetKey()
+
+		foundObject := tk.GetObject()
+		foundObjectType, _ := tuple.SplitObject(foundObject)
+
+		if _, ok := foundObjectsMap.Load(foundObject); ok {
+			// todo(jon-whit): we could optimize this by avoiding reading this
+			// from the database in the first place
+
+			// if we've already evaluated/found the object, then continue
+			continue
+		}
+
+		if foundObjectType == sourceObjectType {
+			if foundCount != nil && atomic.AddUint32(foundCount, 1) > req.limit {
+				break
+			}
+
+			resultChan <- foundObject
+			foundObjectsMap.Store(foundObject, struct{}{})
+		}
+
+		g.Go(func() error {
+			return c.streamedConnectedObjects(ctx, &ConnectedObjectsRequest{
+				StoreID:          store,
+				ObjectType:       sourceObjectType,
+				Relation:         sourceObjectRel,
+				User:             &openfgapb.ObjectRelation{Object: foundObject, Relation: req.targetUserRef.Relation},
+				ContextualTuples: req.contextualTuples,
```

Comment:
isn't `Limit:            req.limit,` missing here?

Reply:
Resolved by addressing the other recommendation.